### ISSUE TITLE
data given to loss layer is not mean subtracted

### ISSUE THREAD

**Issue by erogol** [BVLC/caffe#1805]
As I train an AE , I observed that the reconstructed images are pretty similar to given data without deprocessing in python interface. However I transform the train data at the beginning by subtracting pixel-wise mean value. Therefore I expect that the reconstructed image needs to be summed with the mean matrix before proper visualization but this is not the case. I believe this is because the loss layer does not get the transformed image, instead it takes the raw image that I give.

**Comment by shelhamer** [comment]
@erogol this is by design -- there is no reason to couple a loss to a given task like this. Simply mean subtract your ground truth for training. That is, give the loss the target you want it to learn.

Once again, please discuss use cases on [caffe-users](https://groups.google.com/forum/#!forum/caffe-users) mailing list.

**Comment by erogol** [comment]
Even I am not agreeing on  the "design choice", at least I see the reason. Then all the ground_thruth values need to be preprocessed beforehand and don't use data_layer transformations in the training tim for ground truth preprocessing. I guess this is not very good for AE like applications and proper for a data_layer expectations of the users.

**Comment by shelhamer** [comment]
@eregol the goal is generality -- consider that Euclidean loss and Sigmoid Cross-Entropy should work for any targets in their ranges of [-Inf, +Inf] and [0, 1] respectively.

The special cases, like AE reconstruction, can be handled perfectly well -- here the input and ground truth are exactly the same, so just re-use the same blob for both.

**Comment by erogol** [comment]
Ok here is the problem I encounter with my prototxt file. Maybe it will be more clear.

name: "BlaNet"
layers {
  name: "data"
  type: DATA
  top: "data"
  top: "label"
  data_param {
    source: "Imgs_train.leveldb"
    backend: LEVELDB
    batch_size: 150
  }
  transform_param {
    crop_size: 195
    mean_file: "mean_file.binaryproto"
    mirror: true  
    #scale: 0.00390625  
  }
  include: { phase: TRAIN }   
}

layers {
  name: "data"
  type: DATA
  top: "data"
  top: "label"
  data_param {
    source: "Imgs_test.leveldb"
    backend: LEVELDB
    batch_size: 150
  }
  transform_param {
    crop_size: 195
    mean_file: "mean_file.binaryproto"
    mirror: false
    #scale: 0.00390625
  }
  include: { phase: TEST }
}

layers {
  name: "SILENCE"
  type : SILENCE
  bottom : "label"
}

layers {
  name: "dummy_data"
  type: DUMMY_DATA
  dummy_data_param{
    data_filler{
      type: "gaussian"
      std: 2.5
      mean: 0
    }
    num: 150
    channels : 3
    height : 195
    width : 195
  }
  top: "noise"
}

layers {
  name: "add_noise"
  type: ELTWISE
  bottom: "data"
  bottom: "noise"
  top: "data_noised"
  eltwise_param {
    operation: SUM
  }
}

layers {
  name: "conv1"
  type: CONVOLUTION
  bottom: "data_noised"
  top: "conv1"
  blobs_lr: 1
  blobs_lr: 2
  weight_decay: 1
  weight_decay: 0
  convolution_param {
    num_output: 96
    kernel_size: 11
    stride: 4
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}

layers {
  name: "relu1"
  type: RELU
  bottom: "conv1"
  top: "conv1"
}

layers {
  name: "deconv1"
  type: DECONVOLUTION
  bottom: "conv1"
  top: "deconv1"
  blobs_lr: 1
  blobs_lr: 2
  weight_decay: 1
  weight_decay: 0
  convolution_param {
    num_output: 3
    pad: 0
    kernel_size: 11
    stride: 4
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }
}

layers {
  name: "relu2"
  type: RELU
  bottom: "deconv1"
  top: "deconv1"
}

layers {
  bottom: "deconv1"
  bottom: "data_noised"
  top: "l2_error"
  name: "loss"
  type: EUCLIDEAN_LOSS
  loss_weight: 1
}


The thing here is, even I give the "data_noised" to loss layer, still the "data" is propagated to loss layer internally or at least the non-mean subtracted version of the data_noised. I know this since after training, if I give a mean-subtracted image, I see fully-reconstructed image at the deconvolution layer without the need of mean correction. Is this still valid situation?
